### PR TITLE
adding env var handling

### PR DESCRIPTION
**Pull request by nitro-neal**
# Overview
Adding environment variables that get loading into the ssi-service config on startup

# Description
This change refactors the LoadConfig and adds the mutation of the config in seperate functions.
checkValidConfigPath - This checks if a config path is provided and if a valid config is at the path
parseAndApplyDefaults - Creates a new config object and assigns the base defaults
loadDefaultServicesConfig - If there is no toml file we will programmatically create the config
loadTomlFileConfig - if there is a toml file present we will load in the toml config
applyEnvVariables - finally after config is loaded if there are environment variables we will add these to the config

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -9,17 +9,21 @@ import (
...
 	"github.com/BurntSushi/toml"
 	"github.com/ardanlabs/conf"
+	"github.com/joho/godotenv"
 	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 )
 
 const (
 	DefaultConfigPath = "config/config.toml"
+	DefaultEnvPath    = "config/.env"
```
Comment (decentralgabe) [inline config/config.go:19]: does this pick up every `*.env` file in the directory?

File: config/config.go [hunk 2]
```diff
@@ -9,17 +9,21 @@ import (
...
 	ConfigFileName    = "config.toml"
 	ServiceName       = "ssi-service"
 	ConfigExtension   = ".toml"
 
-	DefaultServiceEndpoint = "http://localhost:8080"
+	DefaultServiceEndpoint      = "http://localhost:8080"
+	EnvVariableKeystorePassword = "KEYSTORE_PASSWORD"
```
Comment (decentralgabe) [inline config/config.go:25]: may want to make a `type EnvironmentVariable string` for these

File: config/config.go [hunk 3]
```diff
@@ -155,84 +159,155 @@ func (p *PresentationServiceConfig) IsEmpty() bool {
...
+	}
+
+	// create the config object
+	var config SSIServiceConfig
+
+	if err := parseAndApplyDefaults(config); err != nil {
+		return nil, errors.Wrap(err, "parse and apply defaults")
+	}
+
+	if loadDefaultConfig == true {
```
Comment (decentralgabe) [inline config/config.go:174]:
```suggestion
	if loadDefaultConfig {
```

File: config/config.go [hunk 4]
```diff
@@ -155,84 +159,155 @@ func (p *PresentationServiceConfig) IsEmpty() bool {
...
+		loadDefaultServicesConfig(&config)
+	} else {
+		if err := loadTomlFileConfig(path, &config); err != nil {
+			return nil, errors.Wrap(err, "load toml config ")
```
Comment (decentralgabe) [inline config/config.go:178]: extra space

File: config/config.go [hunk 5]
```diff
@@ -155,84 +159,155 @@ func (p *PresentationServiceConfig) IsEmpty() bool {
...
+		},
+		IssuingServiceConfig: IssuingServiceConfig{
+			BaseServiceConfig: &BaseServiceConfig{Name: "issuing"},
+		},
+	}
+
+	config.Services = servicesConfig
+}
+
+func loadTomlFileConfig(path string, config *SSIServiceConfig) error {
```
Comment (decentralgabe) [inline config/config.go:264]:
```suggestion
func loadTOMLConfig(path string, config *SSIServiceConfig) error {
```

File: config/config.go [hunk 6]
```diff
@@ -155,84 +159,155 @@ func (p *PresentationServiceConfig) IsEmpty() bool {
...
+	if present {
+		if config.Services.StorageOption == nil {
+			config.Services.StorageOption = make(map[string]interface{})
 		}
 
-		// apply defaults if not included in toml file
-		if config.Services.CredentialConfig.BaseServiceConfig.ServiceEndpoint == "" {
-			config.Services.CredentialConfig.BaseServiceConfig.ServiceEndpoint = config.Services.ServiceEndpoint
+		storageOptionMap, ok := config.Services.StorageOption.(map[string]interface{})
+		if ok == false {
```
Comment (decentralgabe) [inline config/config.go:304]:
```suggestion
		if !ok {
```

**Review by decentralgabe — COMMENTED** [review]
looks good - small comments. may be worth adding a section to the readme on config and the order in which it is loaded

File: config/config.go (hunk 6 above)
Comment (andresuribe87) [inline config/config.go:304]:
```suggestion
		if !ok {
```

File: config/config.go [hunk 7]
```diff
@@ -155,84 +159,155 @@ func (p *PresentationServiceConfig) IsEmpty() bool {
...
+			return errors.New("storage option casting")
```
Comment (andresuribe87) [inline config/config.go:305]:
```suggestion
			return errors.New("storage option must be of type map[string]interface{}")
```

File: config/.env.example [hunk 8]
```diff
@@ -0,0 +1,2 @@
+KEYSTORE_PASSWORD=default-keystore-password
+DB_PASSWORD=default-db-password
```
Comment (andresuribe87) [inline config/.env.example:2]: This is awesome! Would you mind adding instructions to the README?

**Review by andresuribe87 — APPROVED** [review]
Looks good! Couple small suggestions.

File: config/config.go (hunk 1 above)
Comment (nitro-neal) [reply config/config.go:19]: no, just a file named .env in the config/ dir

File: config/config.go (hunk 2 above)
Comment (nitro-neal) [reply config/config.go:25]: Yea that makes sense

File: config/.env.example (hunk 8 above)
Comment (nitro-neal) [reply config/.env.example:2]: yea good call

**Pull request merged**